### PR TITLE
When timestamp is untrusted, display the primary signature details and modified output for various verbosity levels in nuget verify command

### LINKED ISSUE #10316: dotnet nuget verify is too quiet
## Details about Problem

NuGet product used: dotnet.exe

dotnet.exe --version (if appropriate): 5.0.100

## Detailed repro steps so we can see the same problem

`dotnet nuget verify` does not output anything by default:

```
PS > dotnet nuget verify --all .\newtonsoft.json.12.0.3.nupkg
PS > dotnet nuget verify .\newtonsoft.json.12.0.3.nupkg
```

No output is given at all for these commands. This is unfortunate as the `verify` command's output provides useful information when configuring `trusted-signers` (see this [blog post](https://devblogs.microsoft.com/nuget/lock-down-your-dependencies-using-configurable-trust-policies/#configure-trusted-package-authors)).

However, command output is shown if I increase the verbosity to `normal`:

```
PS > dotnet nuget verify .\newtonsoft.json.12.0.3.nupkg --verbosity normal

Verifying Newtonsoft.Json.12.0.3
.\newtonsoft.json.12.0.3.nupkg
Signature Hash Algorithm: SHA256
Timestamp: 11/8/2019 4:56:46 PM
Verifying author primary signature's timestamp with timestamping service certificate:
  Subject Name: CN=TIMESTAMP-SHA256-2019-10-15, O="DigiCert, Inc.", C=US
  SHA1 hash: 0325BD505EDA96302DC22F4FA01E4C28BE2834C5
  SHA256 hash: 481F4373272D98586C5364B6C115E82425675AEBFD9FACF7ADC464FA2FFFB8F0
  Issued by: CN=DigiCert SHA2 Assured ID Timestamping CA, OU=www.digicert.com, O=DigiCert Inc, C=US
  Valid from: 9/30/2019 5:00:00 PM to 10/16/2030 5:00:00 PM
Signature type: Author
Verifying the author primary signature with certificate:
  Subject Name: CN=Json.NET (.NET Foundation), O=Json.NET (.NET Foundation), L=Redmond, S=wa, C=US, SERIALNUMBER=603 389 068
  SHA1 hash: 4CFB89FAA49539A58968D81960B3C1258E8F6A34
  SHA256 hash: A3AF7AF11EBB8EF729D2D91548509717E7E0FF55A129ABC3AEAA8A6940267641
  Issued by: CN=.NET Foundation Projects Code Signing CA, O=.NET Foundation, C=US
  Valid from: 10/24/2018 5:00:00 PM to 10/29/2021 5:00:00 AM
Timestamp: 11/8/2019 5:28:02 PM
Verifying repository countersignature's timestamp with timestamping service certificate:
  Subject Name: CN=Symantec SHA256 TimeStamping Signer - G3, OU=Symantec Trust Network, O=Symantec Corporation, C=US
  SHA1 hash: A9A4121063D71D48E8529A4681DE803E3E7954B0
  SHA256 hash: C474CE76007D02394E0DA5E4DE7C14C680F9E282013CFEF653EF5DB71FDF61F8
  Issued by: CN=Symantec SHA256 TimeStamping CA, OU=Symantec Trust Network, O=Symantec Corporation, C=US
  Valid from: 12/22/2017 4:00:00 PM to 3/22/2029 4:59:59 PM
Signature type: Repository
nuget-v3-service-index-url: https://api.nuget.org/v3/index.json
nuget-package-owners: jamesnk, newtonsoft
Verifying the repository countersignature with certificate:
  Subject Name: CN=NuGet.org Repository by Microsoft, O=NuGet.org Repository by Microsoft, L=Redmond, S=Washington, C=US
  SHA1 hash: 8FB6D7FCF7AD49EB774446EFE778B33365BB7BFB
  SHA256 hash: 0E5F38F57DC1BCC806D8494F4F90FBCEDD988B46760709CBEEC6F4219AA6157D
  Issued by: CN=DigiCert SHA2 Assured ID Code Signing CA, OU=www.digicert.com, O=DigiCert Inc, C=US
  Valid from: 4/9/2018 5:00:00 PM to 4/14/2021 5:00:00 AM

Successfully verified package 'Newtonsoft.Json.12.0.3'.
```

### LINKED ISSUE #10535: Signing:  primary signature details not displayed when certificate expired and timestamp untrusted
NuGet product used (NuGet.exe | Visual Studio | MSBuild.exe | dotnet.exe):  NuGet.exe

Product version:  5.8.1.7021

Worked before?  No.  This reproes as early as 4.9.4.5839 and probably earlier (but did not verify).

## Repro steps and/or sample project

1.  Extract the contents of [AuthorExpired.1.0.0.zip](https://github.com/NuGet/Home/files/5922200/AuthorExpired.1.0.0.zip) to your local drive.
2.  Add AuthorExpired.1.0.0.nupkg.certificates\0.cer as a trusted root authority.
3.  In a command prompt execute:
```
nuget.exe verify -all AuthorExpired.1.0.0.nupkg
```
**Note:  be sure to remove the trusted root authority added in step 2.**

### Expected results:
I expect to see:
*  primary signature details
*  a warning that the primary signature's certificate has expired
*  timestamp signature details
*  a warning or error that the primary signature's timestamp signature's certificate is untrusted

### Actual results:
I see:
*  timestamp signature details
*  an error that the primary signature's timestamp signature's certificate is untrusted
```
Verifying AuthorExpired.1.0.0
E:\Trash\ClassLibrary3dfsdsf\packageSource\AuthorExpired.1.0.0\AuthorExpired.1.0.0.nupkg

Signature Hash Algorithm: SHA256
Timestamp: 2/3/2021 3:36:12 PM

Verifying author primary signature's timestamp with timestamping service certificate:
  Subject Name: CN=NuGet Test Root Certificate Authority (40998d55-3d73-4a3b-a689-55e30c1fac3c), O=NuGet, L=Redmond, S=WA, C=US
  SHA1 hash: 6B2378A3DC9CA185252BB66F24F262D129165B5B
  SHA256 hash: 61B18DE3D814FA7960C6ED62DB20BEA6D0F8D65F678464D7D7C9227E7D5DEFBD
  Issued by: CN=NuGet Test Root Certificate Authority (40998d55-3d73-4a3b-a689-55e30c1fac3c), O=NuGet, L=Redmond, S=WA, C=US
  Valid from: 2/3/2021 3:36:11 PM to 12/31/2099 4:00:00 PM

NU3028: The author primary signature's timestamp found a chain building issue: UntrustedRoot: A certificate chain processed, but terminated in a root certificate which is not trusted by the trust provider.
Finished with 1 errors and 0 warnings.

Package signature validation failed.
```

### PR DESCRIPTION
**Pull request by kartheekp-ms**
## Bug

Fixes: https://github.com/nuget/home/issues/10316
Fixes: https://github.com/nuget/home/issues/10535

Regression? Last working version:

## Description

Modified output displayed for various verbosity levels of `dotnet nuget verify` and `nuget.exe verify` command as per [spec](https://github.com/NuGet/Home/blob/dev-kartheekp-ms-verify-verbosity/designs/Verify-Verbosity.md#solution). Improved log messages & error codes in docs to improve the customer experience based upon the learnings from the recent Debian incident.

When timestamp is untrusted, display the primary signature details to fix https://github.com/nuget/home/issues/10535

Once this PR is merged, `dotnet nuget verify` and `nuget.exe verify` command displays output in the following order.

- primary signature details
- timestamp signature details
- warnings or errors

## PR Checklist

- [x] PR has a meaningful title
- [x] PR has a linked issue.
- [x] Described changes

- **Tests**
  - [x] Automated tests added
  - **OR**
  
  - [ ] Test exception
  - **OR**
  - [ ] N/A 

- **Documentation**
  
  - [x] Documentation PR or issue filled
    - https://github.com/nuget/docs.microsoft.com-nuget/issues/2359
    - https://github.com/nuget/docs.microsoft.com-nuget/issues/2346
    - https://github.com/nuget/docs.microsoft.com-nuget/issues/2347
    - https://github.com/nuget/docs.microsoft.com-nuget/issues/2345
  - **OR**
  - [ ] N/A

### CODE REVIEW

File: src/NuGet.Core/NuGet.Packaging/PackageArchiveReader.cs [hunk 1]
```diff
@@ -374,21 +381,27 @@ public override Task<bool> IsSignedAsync(CancellationToken token)
 
             ThrowIfZipReadStreamIsNull();
 
-            var isSigned = false;
-
 #if IS_SIGNING_SUPPORTED
-            using (var zip = new ZipArchive(ZipReadStream, ZipArchiveMode.Read, leaveOpen: true))
+            if (!_isSigned.HasValue)
```
Comment (kartheekp-ms) [inline src/NuGet.Core/NuGet.Packaging/PackageArchiveReader.cs:385]: When debugging `dotnet nuget verify` command, I noticed that `IsSignedAsync` method is invoked multiple times in the same code path. Hence added a nullable bool `_isSigned` to avoid executing the logic multiple times (a minor perf. improvement).

File: src/NuGet.Core/NuGet.Packaging.Extraction/Strings.resx [hunk 2]
```diff
@@ -341,7 +341,7 @@ Valid from:</comment>
     <value>The package hash uses an unsupported hash algorithm.</value>
   </data>
   <data name="SignaturePackageIntegrityFailure" xml:space="preserve">
-    <value>The package integrity check failed.</value>
+    <value>The package integrity check failed. The package has changed since it was signed. Try clearing the local cache and run nuget operation again.</value>
```
Comment (nkolev92) [inline src/NuGet.Core/NuGet.Packaging.Extraction/Strings.resx]: I think the term local cache is ambiguous. 

Are we referring to the http cache? global packages folder? 

Note the specific language in https://docs.microsoft.com/en-us/nuget/consume-packages/managing-the-global-packages-and-cache-folders

Comment (kartheekp-ms) [reply src/NuGet.Core/NuGet.Packaging.Extraction/Strings.resx]: Great question Nikolche. I have not thought about that. I produced this string based on content in https://docs.microsoft.com/en-us/nuget/reference/errors-and-warnings/nu3008#solution. I will take a step back and learn more about it.

Comment (kartheekp-ms) [reply src/NuGet.Core/NuGet.Packaging.Extraction/Strings.resx]: Reading from the [package-installation-process doc](https://docs.microsoft.com/en-us/nuget/concepts/package-installation-process), packages are searched in the following order during install.

 `global packages folder` -> `fallback folders` -> `local sources` -> `http-cache` ->`online sources`.

My understanding is when a package is fetched from an online source it will be cached in the `http-cache` first and then perform `package signature validation` before saving a copy in the `global packages folder`. If the package fails signature validation because the contents have been changed since it was signed, then the package will not be saved to the global packages folder but remain in the `http-cache`. When we try to install again the same package before http-cache gets invalidated, then the package will be fetched from the `http-cache`. Hence user may get the same warning again. 

Given the above details, I would like to change the warning message to `The package integrity checks failed. The package has changed since it was signed. Try clearing the local http-cache and run nuget operation again.`

Comment (nkolev92) [reply src/NuGet.Core/NuGet.Packaging.Extraction/Strings.resx]: Right on all accounts :) 
Great job, I agree with your proposal.

File: src/NuGet.Core/NuGet.Commands/VerifyCommand/VerifyCommandRunner.cs [hunk 3]
```diff
@@ -92,16 +92,21 @@ private async Task<int> VerifySignatureForPackageAsync(string packagePath, ILogg
...
+                logger.LogMinimal(Environment.NewLine + string.Format(CultureInfo.CurrentCulture,
                     Strings.VerifyCommand_VerifyingPackage,
                     packageIdentity.ToString()));
                 logger.LogInformation($"{packagePath}{Environment.NewLine}");
 
                 var logMessages = verificationResult.Results.SelectMany(p => p.Issues).ToList();
-                await logger.LogMessagesAsync(logMessages);
+
+                //log issues first
+                IEnumerable<SignatureLog> issues = logMessages.Where(m => m.Level < LogLevel.Warning);
```
Comment (nkolev92) [inline src/NuGet.Core/NuGet.Commands/VerifyCommand/VerifyCommandRunner.cs]: nit: 
Multiple assignments to the same var can sometimes be error prone when coding. 

You can consider directlly calling logger.LogMessagesAsync on the expression.

File: src/NuGet.Core/NuGet.Packaging/PackageArchiveReader.cs (hunk 1 above)
Comment (nkolev92) [reply src/NuGet.Core/NuGet.Packaging/PackageArchiveReader.cs:385]: nit: Consider positive checks where possible.

**Review by nkolev92 — APPROVED** [review]
👏

File: src/NuGet.Core/NuGet.Packaging/Signing/Verification/SignatureTrustAndValidityVerificationProvider.cs [hunk 4]
```diff
@@ -219,28 +219,28 @@ public SignatureTrustAndValidityVerificationProvider(IEnumerable<KeyValuePair<st
...
-                certificateExtraStore);
-
             return new SignatureVerificationSummary(
                 status.SignatureType,
                 status.Status,
                 status.Flags,
                 status.Timestamp,
                 status.ExpirationTime,
-                timestampSummary.Issues.Concat(status.Issues));
+                status.Issues.Concat(timestampSummary.Issues));
```
Comment (heng-liu) [inline src/NuGet.Core/NuGet.Packaging/Signing/Verification/SignatureTrustAndValidityVerificationProvider.cs:243]: May I know if those changes are for fixing https://github.com/NuGet/Home/issues/10535?
I understand that we don't have test infrastructure for untrustedroot for timestamp, so the test is not added.
May I know if you verified in other ways (e.g. manually patch and test) to make sure https://github.com/NuGet/Home/issues/10535 is fixed? Thanks.

**Review by heng-liu — APPROVED** [review]
Good job @kartheekp-ms !
Just have one question in the below comment.

File: src/NuGet.Core/NuGet.Packaging/Signing/Verification/SignatureTrustAndValidityVerificationProvider.cs (hunk 4 above)
Comment (kartheekp-ms) [reply src/NuGet.Core/NuGet.Packaging/Signing/Verification/SignatureTrustAndValidityVerificationProvider.cs:243]: > May I know if those changes are for fixing NuGet/Home#10535?

Yes, that is correct.  I verified this fix by modifying the asserts statements in the following tests.
     
- `SignatureTrustAndValidityVerificationProviderTests.GetTrustResultAsync_WithNoIgnoringTimestamp_TimestampWithGeneralizedTimeOutsideCertificateValidity_FailAsync()` https://github.com/NuGet/NuGet.Client/pull/3992/files#diff-ac0c57f69076c8fd3f9e38cdbcca84fa2de8e9052cbe91d9fac7e108f57d2b84R556
- `PackageSignatureVerifierTests.VerifySignaturesAsync_WithTimestampChainingToUntrustedRoot_NotAllowIgnoreTimestamp_FailAsync()` https://github.com/NuGet/NuGet.Client/pull/3992/files#diff-6a15bb307e86dc7012938479f8a3fe9aff77788c02854afca190d0a2e5b4e405R768

I have also verified this fix manually patching dotnet SDK.

Comment (heng-liu) [reply src/NuGet.Core/NuGet.Packaging/Signing/Verification/SignatureTrustAndValidityVerificationProvider.cs:243]: That's great! 
Thanks for correcting me.
Just found the setting `AllowIgnoreTimestamp` for VerifyCommandDefaultPolicy is set to false.
And the two test cases set `AllowIgnoreTimestamp` to false.
So the two test cases could repro the same issue with NuGet/Home#10535.

File: src/NuGet.Core/NuGet.Commands/VerifyCommand/VerifyCommandRunner.cs (hunk 3 above)
Comment (kartheekp-ms) [reply src/NuGet.Core/NuGet.Commands/VerifyCommand/VerifyCommandRunner.cs]: Addressed this feedback in the latest commit.

File: src/NuGet.Core/NuGet.Packaging/PackageArchiveReader.cs (hunk 1 above)
Comment (kartheekp-ms) [reply src/NuGet.Core/NuGet.Packaging/PackageArchiveReader.cs:385]: I am leaning towards keeping the current implementation as is. If you prefer a positive check, I can change the code as shown below. Please let me know your feedback.

``` cs
            if (_isSigned.HasValue)
                return Task.FromResult(_isSigned.Value);

            _isSigned = false;

            using (var zip = new ZipArchive(ZipReadStream, ZipArchiveMode.Read, leaveOpen: true))
            {
                var signatureEntry = zip.GetEntry(SigningSpecifications.SignaturePath);

                if (signatureEntry != null &&
                   string.Equals(signatureEntry.Name, SigningSpecifications.SignaturePath, StringComparison.Ordinal))
                {
                    _isSigned = true;
                }
            }

            return Task.FromResult(_isSigned.Value);
```

File: src/NuGet.Core/NuGet.Commands/VerifyCommand/VerifyCommandRunner.cs [hunk 5]
```diff
@@ -92,21 +92,30 @@ private async Task<int> VerifySignatureForPackageAsync(string packagePath, ILogg
...
-                logger.LogInformation(Environment.NewLine + string.Format(CultureInfo.CurrentCulture,
+                logger.LogMinimal(Environment.NewLine + string.Format(CultureInfo.CurrentCulture,
                     Strings.VerifyCommand_VerifyingPackage,
                     packageIdentity.ToString()));
                 logger.LogInformation($"{packagePath}{Environment.NewLine}");
 
                 var logMessages = verificationResult.Results.SelectMany(p => p.Issues).ToList();
-                await logger.LogMessagesAsync(logMessages);
+
+                //log issues first
```
Comment (erdembayar) [inline src/NuGet.Core/NuGet.Commands/VerifyCommand/VerifyCommandRunner.cs:102]: Is this comment correct? You're logging non-issues first.

File: src/NuGet.Core/NuGet.Packaging/Signing/Verification/SignatureTrustAndValidityVerificationProvider.cs [hunk 6]
```diff
@@ -219,28 +219,28 @@ public SignatureTrustAndValidityVerificationProvider(IEnumerable<KeyValuePair<st
             Timestamp timestamp;
             var timestampSummary = GetTimestamp(signature, verifierSettings, out timestamp);
 
+            var status = signature.Verify(
+            timestamp,
```
Comment (erdembayar) [inline src/NuGet.Core/NuGet.Packaging/Signing/Verification/SignatureTrustAndValidityVerificationProvider.cs:223]: Indent here for better readability?

Comment (erdembayar) [reply src/NuGet.Core/NuGet.Packaging/Signing/Verification/SignatureTrustAndValidityVerificationProvider.cs:223]: Also can you put explicit type? Otherwise hard to follow what kind of value is here?

**Review by erdembayar — COMMENTED** [review]
Looks good. Just 2 comments.

File: src/NuGet.Core/NuGet.Commands/VerifyCommand/VerifyCommandRunner.cs (hunk 5 above)
Comment (kartheekp-ms) [reply src/NuGet.Core/NuGet.Commands/VerifyCommand/VerifyCommandRunner.cs:102]: yes Erick, the comment is correct. The idea is to log all the issues whose level is `not warning or error` first. If you think it needs to be tweaked, I will create a follow up PR because this PR has passed all checks.

File: src/NuGet.Core/NuGet.Packaging/Signing/Verification/SignatureTrustAndValidityVerificationProvider.cs (hunk 6 above)
Comment (kartheekp-ms) [reply src/NuGet.Core/NuGet.Packaging/Signing/Verification/SignatureTrustAndValidityVerificationProvider.cs:223]: If you are okay, I will create a follow up PR addressing above comments because this PR has passed all checks.

File: src/NuGet.Core/NuGet.Commands/VerifyCommand/VerifyCommandRunner.cs (hunk 5 above)
Comment (erdembayar) [reply src/NuGet.Core/NuGet.Commands/VerifyCommand/VerifyCommandRunner.cs:102]: Ok. I thought `issue` means something has problem such as warnings and errors.
I believe correct wording would be '//log non-issues first'.
Of you can address in another PR.

Comment (kartheekp-ms) [reply src/NuGet.Core/NuGet.Commands/VerifyCommand/VerifyCommandRunner.cs:102]: I can address this comment in this PR, but my only concern is it may take 2 or 3 attempts for a green build.

**Review by erdembayar — APPROVED** [review]
Good job.

File: src/NuGet.Core/NuGet.Commands/VerifyCommand/VerifyCommandRunner.cs (hunk 5 above)
Comment (erdembayar) [reply src/NuGet.Core/NuGet.Commands/VerifyCommand/VerifyCommandRunner.cs:102]: Ok. I approved this PR.
